auth/dfa.js

```diff
@@ -0,0 +1,348 @@
+  });
+};
+
+// Export sortRolesForDFAGeneration
+exports.sortRolesForDFAGeneration = sortRolesForDFAGeneration;
+
+/**
+ * Compare socpes a and b to see which comes first if sorted
+ * Such that 'a*' comes before 'a', but otherwise normal order.
+ */
+let scopeCompare = (a, b) => {
+  let n = a.length;
+  let m = b.length;
+
+  let d = Math.abs(n - m);
+  if (d == 0) {
+    if (a[n - 1] === '*' || b[n - 1] === '*') {
+      if (a.startsWith(b.slice(0, -1))) {
+        if (a[n - 1] === '*') {
+          return -1;
+        }
+        if (b[n - 1] === '*') {
+          return 1;
+        }
+      }
+    }
+  } else if (d == 1) {
+    if (n > m && a[n - 1] === '*') {
+      if (a.startsWith(b)) {
+        return -1;
+      }
+    } else if (b[n - 1] === '*') {
+      if (b.startsWith(a)) {
+        return 1;
+      }
+    }
+  }
+
+  return a < b ? -1 : 1;
+};
+
+/** Assumes scopes to be unique, and sorts scopes for use with mergeScopeSets */
+let sortScopesForMerge = (scopes) => {
+  return scopes.sort(scopeCompare);
+};
+
+// Export sortScopesForMerge
+exports.sortScopesForMerge = sortScopesForMerge;
+
+/**
+ * Take two sets of sorted scopes and merge them removing duplicates,
+ * as well as scopes implied by a star-scopes.
+ */
+let mergeScopeSets = (scopes1, scopes2) => {
+  let n = scopes1.length;
+  let m = scopes2.length;
+  let i = 0;
+  let j = 0;
+  let scopes = [];
+  while (i < n && j < m) {
+    let s1 = scopes1[i];
+    let s2 = scopes2[j];
+    let scope = null;
+    if (s1 === s2) {
+      scopes.push(s1);
+      scope = s1;
+      i += 1;
+      j += 1;
+    } else {
+      let z = scopeCompare(s1, s2);
+      if (z < 0) {
+        scope = s1;
+        scopes.push(s1);
+        i += 1;
+      } else {
+        scope = s2;
+        scopes.push(s2);
+        j += 1;
+      }
+    }
+    // If we just added a star scope, we have to skip everything that matches
+    if (scope.endsWith('*')) {
+      let prefix = scope.slice(0, -1);
+      while(i < n && scopes1[i].startsWith(prefix)) {
+        i += 1;
+      }
+      while(j < m && scopes2[j].startsWith(prefix)) {
+        j += 1;
+      }
+    }
+  }
+  while (i < n) {
+    let scope = scopes1[i];
+    scopes.push(scope);
+    i += 1;
+    if (scope.endsWith('*')) {
+      let prefix = scope.slice(0, -1);
+      while(i < n && scopes1[i].startsWith(prefix)) {
+        i += 1;
+      }
+    }
+  }
+  while (j < m) {
+    let scope = scopes2[j];
+    scopes.push(scope);
+    j += 1;
+    if (scope.endsWith('*')) {
+      let prefix = scope.slice(0, -1);
+      while(j < m && scopes2[j].startsWith(prefix)) {
+        j += 1;
+      }
+    }
+  }
+  return scopes;
+};
+
+// Export mergeScopeSets
+exports.mergeScopeSets = mergeScopeSets;
+
+/**
+ * Build a DFA of states on the form:
+ * ```js
+ * {
+ *   'a': // next state if the current character is 'a'
+ *   'end': [] // set of roles granted if the scope ends here
+ *   'prefix': {...} // role granted if the scope matches so far
```

Comment:
This needs to be expanded out a bit.  It sounds like the states are nested, but what's the difference between "end" and "prefix", and why is the latter an object?